lib/utils/platform/iot_clock_afr.c

```diff
@@ -0,0 +1,215 @@
+#define _MILLISECONDS_PER_TICK          ( _MILLISECONDS_PER_SECOND / configTICK_RATE_HZ ) /**< Milliseconds per FreeRTOS tick. */
+
+/*-----------------------------------------------------------*/
+
+
+/*-----------------------------------------------------------*/
+
+/*  Private Callback function for timer expiry, delegate work to a Task to free 
+    up the timer task for managing other timers */
+static void prvTimerCallback( TimerHandle_t xTimerHandle )
+{
+    _timerInfo_t * pxTimer = ( _timerInfo_t * ) pvTimerGetTimerID( xTimerHandle );
+
+    /* The value of the timer ID, set in timer_create, should not be NULL. */
+    configASSERT( pxTimer != NULL );
+
+    /* Restart the timer if it is periodic. */
+    if( pxTimer->xTimerPeriod > 0 )
+    {
+        xTimerChangePeriod( xTimerHandle, pxTimer->xTimerPeriod, 0 );
+    }
+
+    /* Call timer Callback from this task */
+    pxTimer->threadRoutine( ( void * ) pxTimer->pArgument);
+}
+
+/*-----------------------------------------------------------*/
+
+bool IotClock_GetTimestring( char * const pBuffer,
+                                size_t bufferSize,
+                                size_t * const pTimestringLength )
+{
+    uint64_t milliSeconds = IotClock_GetTimeMs( );
+    int timestringLength = 0;
+
+    /* Convert the localTime struct to a string. */
+    timestringLength = snprintf( pBuffer, bufferSize, "%llu",  milliSeconds );
+
+    /* Check for error from no string */
+    if( timestringLength == 0 )
+    {
+        return false;
+    }
+
+    /* Set the output parameter. */
+    *pTimestringLength = timestringLength;
+
+    return true;
+}
+
+/*-----------------------------------------------------------*/
+
+uint64_t IotClock_GetTimeMs( void )
+{
+    TimeOut_t xCurrentTime = { 0 };
+
+    /* Intermediate variable used to convert TimeOut_t to struct timespec.
+     * Also used to detect overflow issues. It must be unsigned because the
+     * behavior of signed integer overflow is undefined. */
+    uint64_t ullTickCount = 0ULL;
+
+    /* Get the current tick count and overflow count. vTaskSetTimeOutState()
+        * is used to get these values because they are both static in tasks.c. */
+    vTaskSetTimeOutState( &xCurrentTime );
+
+    /* Adjust the tick count for the number of times a TickType_t has overflowed. */
+    ullTickCount = ( uint64_t ) ( xCurrentTime.xOverflowCount ) << ( sizeof( TickType_t ) * 8 );
+
+    /* Add the current tick count. */
+    ullTickCount += xCurrentTime.xTimeOnEntering;
+
+    /* Return the ticks converted to Milliseconds */
+    return ullTickCount * _MILLISECONDS_PER_TICK;
+}
+
+/*-----------------------------------------------------------*/
+
+bool IotClock_TimerCreate( IotTimer_t * const pNewTimer,
+                           IotThreadRoutine_t expirationRoutine,
+                           void * pArgument )
+{
+    _timerInfo_t * pxTimer = ( _timerInfo_t*) pNewTimer;
```

Comment:
Check parameters? or Assert? (this comment applies to all public APIs, so i will not repeat it)

Reply:
Agreed.